docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+* [Strategies for Returning References in Rust](http://bryce.fisher-fleig.org/blog/strategies-for-returning-references-in-rust/index.html)
+  * Though I don't believe we should be doing this, it's still an interesting read.
+
+## Other Useful Links
+
+* Write and run Rust snippets: https://play.rust-lang.org
+  * You can save snippets in permantent links and incluce them as working examples in docs.
+* Write code and see what it compiles to: https://rust.godbolt.org
+  * Note that you need to enable compiler optimizations via `-C opt-level=3` if you are looking to meaningfully compare two code snippets.
+
+# Naming Conventions
+
+### No Special Naming :: required
+
+Follow the naming advice in [Naming - Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html) at all times, and use [Rust RFC 0430](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md) for any ambiguous issues (e.g. Crate naming).
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
+
+* Infer Crate names from package names where feasible:
+  * A package is the directory in which the Cargo.toml exists.
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with underscore__ (a.k.a "snake_case") to split words in package names if needed
+
+I believe this would result in Crate names (which must be Rust identifiers) being identical to the package name at all times.
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
+
+**Open Question**: Should we have code in the top-level module of a Crate or always use named sub-modules? This probably ties into what Crates we intend to publish and the granularity of the APIs.
+
+## Shared Internal Code
+
+Rust supports having non-public sub-modules within a Crate, and only the top-level "implicit" module is exposed by default (other modules can be exported by the top level module). See also [File hierarchy - Rust By Example](https://doc.rust-lang.org/rust-by-example/mod/split.html).
+
+### Use Private Modules :: required
+
+While local private functions can just be defined in any module, shared internal functions should always be implemented in a non-exported sub-module. This avoids any misleading dependencies between "public" modules and should help ensure clean code dependencies. Code in internal modules should still be tested and documented thoroughly. It is perfectly expected that we would have multiple internal modules for different aspects of ICU4X.
```

Comment:
Do we need to do this? `pub(crate)` exists and keeps this stuff reasonably distinct.

Reply:
Happy to remove. I'll add something about pub(crate) etc.